### DISCUSSION TITLE
Need Help with Python Module with Podman.

### DISCUSSION THREAD

**Discussion by kryzhh** [containers/podman#23371] [Q&A] [answered]
Salutations to the one reading this. I am currently working on a project which requires me to use containers to run a program inside them. I built the app just fine (I think). When I try to run it, this is the output:

![image](https://github.com/user-attachments/assets/4bc589ad-64f0-4ee2-adb9-443506be9200)

Podman runs just fine normally.

![image](https://github.com/user-attachments/assets/b5cf3177-e88d-4117-8f48-a31d1d68dba4)

I also have Python module for Docker. And it runs just fine

I tried running ls -l /run/podman/podman.sock and it said permission denied. Then I tried to look for fix, I ran chmod to give access but the output was same (even with root, though it shows output with sudo ls -l /run/podman/podman.sock) 

Here's the Python code:

```py
import docker
import podman
from podman import PodmanClient

from pathlib import Path

global ENGINE, PLATFORM, IMG_NAME
ENGINE = "podman"
PLATFORM = "Linux"
IMG_NAME = "nikto-img" # Normally I just import all 3 variables from config file but for the sake for problem I will just put them directly
# print("This test is just for all Python and Docker based things")
# def read_config():
#     print("Checking for configuration file....")
#     CONF_DIR = Path('.config')

#     if CONF_DIR.is_file(): # If config file is already present, call main function
#         print("Configuration file found!")
#         f = open(".config","r") # Open config file
#         lines = f.readlines() 
#         for i in lines: # Store values from config file into their respective (global) variables
#             global PLATFORM, ENGINE, IMG_NAME 
#             if "OS"in i:
#                 PLATFORM = i.replace("OS: ", '')
#                 PLATFORM = PLATFORM.replace("\n", '')
#             elif "ENGINE" in i:
#                 ENGINE = i.replace("ENGINE: ", '')
#                 ENGINE = ENGINE.replace("\n",'')
#             elif "IMAGE" in i:
#                 IMG_NAME = i.replace("IMAGE: ",  '')
#                 IMG_NAME = IMG_NAME.replace("\n",'')
#         f.close()
#         print(f'OS: {PLATFORM}',f'\nENGINE: {ENGINE}',f'\nIMAGE: {IMG_NAME}')
#     else: # Else ask user to run setup script and exit the app
#         print("Configuration file not present. Run setup script to make a new one.\nExiting app....")
#         quit()

# read_config()
print(f"Testing {ENGINE}.....")

def podman_run(client, image, command): # I took help of AI for this.
    container = client.containers.create(
        image=image,
        command=command,
        remove=True,
        detach=False
    )
    container.start()
    logs = container.logs(stdout=True, stderr=True, stream=False)
    return b''.join(logs).decode('utf-8').strip()


if ENGINE.lower() == "docker": # I have option for user if he got podman or docker 
    Client = docker.from_env()

    log = Client.containers.run(IMG_NAME, "nikto -V")
    log = log.strip().decode('utf-8')
    print(log)
    
    log = Client.containers.run(IMG_NAME, "echo Test successful.")
    log = log.strip().decode('utf-8')
    print(log)

elif ENGINE.lower() == "podman":
    Client = PodmanClient(base_url="unix:///run/podman/podman.sock") # This is where my podman is stored too.

    # Run container with Nikto command
    log = podman_run(Client, f'{IMG_NAME}:latest', ["nikto", "-V"])
    print(log)

    # Run container with test command
    log = podman_run(Client, f'{IMG_NAME}:latest', ["echo", "Test successful."])
    print(log)
```

I am running ParrotOS on VMWare Workstation 17. Here's what I got:

![image](https://github.com/user-attachments/assets/9120c6ea-b9a5-4046-8c0b-fac7d3416cf1)

Any help will be appreciated.

**Comment by rhatdan** [comment]
Are you attempting to connect to a rootful podman socket from a non-root account?  This is not allowed. Docker allows it because of the docker group.  If you want to use rootless podman, you should be connecting to the rootless podman socket, not the rootful one.

**Reply by kryzhh** [reply]
Sorry for the late reply, I even tried running this as root user, it still gave the same error. And since I am new, can you tell me where the rootless podman socket is? Like, what's the base URL?

**Reply by rhatdan** [reply]
Lets take a step back, are you running this on a Linux machine on Mac or Windows?

Could you give the output of `podman info`

**Reply by kryzhh** [reply]
Windows machine but Linux VM
I use Linux as my dev environment

**Reply by rhatdan** [reply, marked as answer]
In rootless mode you can start up the podman.socket service with

systemctl --user start podman.socket

Which should setup podman to listen on 

$XDG_RUNTIME_DIR/podman/podman.sock

**Comment by kryzhh** [comment]
Let me try, will let you know result.
Edit:
![image](https://github.com/user-attachments/assets/e9e29526-600c-4078-b958-0359d2ed5a3f)
what do i put in this url
Also might not reply for a while now cs its gonna be 2am here soon, thanks for your interest so far :D
[host.txt](https://github.com/user-attachments/files/16353757/host.txt)
also heres the podman info you asked for

Edit 2:
I read the file and I think i found the issue. The podman directory I had to use was **/run/user/1000/podman/podman.socket**. While my code dont throw an error now, it is not giving me an output either. Like I run the code and it gives this as output:
![image](https://github.com/user-attachments/assets/94a705f1-4095-4985-9666-957b05083924)
The ideal output is it shows the Nikto version but i guess that's just me mishandling how the output is stored. Thanks for all the help friend. Will mark your next reply as answer :D
